SI-8862 Fix treatment of inherited implicits in package objects

retronym opened the pull request:
Two spots in implicit search fell prey to a trap with package
objects. Members of a package object are entered into the scope
of the enclosing package, but that doesn't form a suitable prefix
for determing the member type.

A REPL transcript paints a picture that speaks a 1000 words:

```
scala> :paste -raw
// Entering paste mode (ctrl-D to finish)

package p { class C[A] { def foo: A = ??? }; object `package` extends C[String] }

// Exiting paste mode, now interpreting.

scala> val p = rootMirror.getPackageIfDefined("p")
warning: there was one deprecation warning; re-run with -deprecation for details
p: $r.intp.global.Symbol = package p

scala> p.info.decls
res0: $r.intp.global.Scope = Scopes(class C, package object p, method foo)

scala> val foo = p.info.decl(TermName("foo"))
foo: $r.intp.global.Symbol = method foo

scala> p.typeOfThis memberType foo
res1: $r.intp.global.Type = => A

scala> val fooOwner = foo.owner
fooOwner: $r.intp.global.Symbol = class C

scala> p.info.decl(nme.PACKAGE).typeOfThis memberType foo
res3: $r.intp.global.Type = => String
```

This commit detects if we find an implicit in a package module,
and then uses the self type of the corresponding package object
as the prefix for the `ImplicitInfo`. This is done in both
`Context.implicitss` (which collects in-scope implicits), and
in `companionImplicitMap` (which harvests them from the implicit
scope.) In addition, it was necessary / possible to remove a special
case that excluded package object implicits, the referenced tests for
SI-3999 now pass without this.

retronym committed:
SI-8862 Fix treatment of inherited implicits in package objects

Two spots in implicit search fell prey to a trap with package
objects. Members of a package object are entered into the scope
of the enclosing package, but that doesn't form a suitable prefix
for determing the member type.

A REPL transcript paints a picture that speaks a 1000 words:

```
scala> :paste -raw
// Entering paste mode (ctrl-D to finish)

package p { class C[A] { def foo: A = ??? }; object `package` extends C[String] }

// Exiting paste mode, now interpreting.

scala> val p = rootMirror.getPackageIfDefined("p")
warning: there was one deprecation warning; re-run with -deprecation for details
p: $r.intp.global.Symbol = package p

scala> p.info.decls
res0: $r.intp.global.Scope = Scopes(class C, package object p, method foo)

scala> val foo = p.info.decl(TermName("foo"))
foo: $r.intp.global.Symbol = method foo

scala> p.typeOfThis memberType foo
res1: $r.intp.global.Type = => A

scala> val fooOwner = foo.owner
fooOwner: $r.intp.global.Symbol = class C

scala> p.info.decl(nme.PACKAGE).typeOfThis memberType foo
res3: $r.intp.global.Type = => String
```

This commit detects if we find an implicit in a package module,
and then uses the self type of the corresponding package object
as the prefix for the `ImplicitInfo`. This is done in both
`Context.implicitss` (which collects in-scope implicits), and
in `companionImplicitMap` (which harvests them from the implicit
scope.) In addition, it was necessary / possible to remove a special
case that excluded package object implicits, the referenced tests for
SI-3999 now pass without this.

retronym commented:
Targetting to 2.12.x as we need to maintain bug compatibility in 2.11.x implicit search. Alternatively I could retarget this with 2.11.x under `-Xsource:2.12`.

retronym commented:
Review by @adriaanm

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Contexts.scala:
```diff
@@ -882,7 +882,13 @@ trait Contexts { self: Analyzer =>
       } else if (owner.isPackageClass) {
         // the corresponding package object may contain implicit members.
-        Some(collectImplicits(owner.tpe.implicitMembers, owner.tpe))
+        owner.tpe.member(nme.PACKAGE) match {
```
Can we factor out the commonalities with `isInPackageObject`? Maybe introduce an `packageOwnerPrefix` method on `Symbol`? Could reuse this to fix a similar mistake (I assume?) in `adapt` in Typers:

```
            val pre =
              if (owner.isPackageClass) owner.thisType
```

retronym replied:
You have just won spot in "Great Moments in Code Review, Package Object edition".

``` scala
package p {
  trait T[X] { def O : { def apply(): X } }
  object `package` extends T[Int] {
    def O: { def apply(): Int } = new { def apply(): Int = 42 }
  }

  object Test {
    def main(args: Array[String]): Unit = {
      val x: Int = O()
    }
  }
}
```

```
% qscalac sandbox/test.scala
sandbox/test.scala:10: error: type mismatch;
 found   : X
 required: Int
      val x: Int = O()
                    ^
one error found
```

retronym replied:
While in the vicinity of `insertApply`

``` scala
      def insertApply(): Tree = {
        assert(!context.inTypeConstructorAllowed, mode) //@M
        val adapted = adaptToName(tree, nme.apply)
        def stabilize0(pre: Type): Tree = stabilize(adapted, pre, MonoQualifierModes, WildcardType)

        // TODO reconcile the overlap between Typers#stablize and TreeGen.stabilize
        val qual = adapted match {
          case This(_) =>
            gen.stabilize(adapted)
          case Ident(_) =>
            val owner = adapted.symbol.owner
            val pre =
              if (owner.isPackageClass) owner.thisType
              else if (owner.isClass) context.enclosingSubClassContext(owner).prefix
              else NoPrefix
            stabilize0(pre)
          case Select(qualqual, _) =>
            stabilize0(qualqual.tpe)
          case other =>
            other
        }
        typedPos(tree.pos, mode, pt) {
          Select(qual setPos tree.pos.makeTransparent, nme.apply)
        }
      }
```

I found the use of `enclosingSubClassContext` suspicious. It was added in ac49199ed. Looking at the old-timey issue it fixed:

http://lrytz.github.io/scala-aladdin-bugtracker/displayItem.do%3Fid=763.html

And then expanding the test case:

``` scala
// Test from http://lrytz.github.io/scala-aladdin-bugtracker/displayItem.do%3Fid=763.html


trait Foo { type T; def apply() : T }
object Test {

  val f : Foo = new Foo { type T = Int; def apply() = 42 }

  def main(args: Array[String]): Unit = {
    val g : Foo = new Foo { type T = Int; def apply() = 42 }
    f()
    g()
  }
}
```

The branch that calls `enclosingSubClassContext` is never executed. If the symbol of the adapted tree is class-owned, we'll always end up with a `Select`, rather than a naked `Ident`. Perhaps this wasn't the case back in the halcyon days of Scala 2.2.

adriaanm replied:
Thanks, oh fearless explorer of the jungle of Paxak Objè! If the fix is amenable to flag-containment, I think it would be great to have under -Xsource:2.12.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -1012,13 +1012,16 @@ trait Implicits {
-                      for (im <- mc.implicitMembers.toList) yield new ImplicitInfo(im.name, singleType(pre, companion), im)
+                      for (im <- mc.implicitMembers.toList) yield new ImplicitInfo(im.name, pre1, im)
                     if (infos.nonEmpty)
-                      infoMap += (sym -> infos)
```
why the deindent?

retronym committed:
More uniform treatment of package objects

  - Introduce `Symbol#packageObject` and `Type#packageObject`
    to lookup the member package object of a package class/module,
    and use this far and wide.
  - Replace the overly complicated (and still buggy) implementation
    of `Context#isInPackageObject` with a one liner. The simplifying
    insight is that if we select a symbol from a package prefix
    that does not own that symbol, it *must* have really been
    selected from the package object.
  - Change implicit search to use the cache in
    `ModuleSymbol#implicitMembers` via `Type#implicitMembers`,
    which lets the client code read more naturally.

Fixes a bug with `adapt::insertApply` that Adriaan spotted in a feat
of lateral thinking. This is tested in t8862b.scala. alladin763.scala
adds the test case from the bug originally remedied by `insertApply`
to check we haven't regressed.

retronym commented:
I had a chance to pull the thread a bit further here: https://github.com/retronym/scala/compare/scala:2.12.x...ticket/8862-2?expand=1

The good news is that it appears we can massively simplify `Contexts#isInPackageObject` by phrasing it instead as `isNotInPackage`.

I'm also introducing `{Type, Symbol}#packageObject` to encapsulate the lookup. I'll leave this PR open for another day or to as my breadcrumb to finish and submit that branch.

adriaanm commented:
That's some seriously exciting WIP you got there, @retronym! Looks like we're finally in reach of a good implementation strategy for package objects.

retronym commented:
The test hangs will be fixed by #4090.

In the meantime, this is ready for review by @adriaanm

retronym commented:
Rebased to get the tests working. @adriaanm I don't feel comfortable moving this change to 2.11.x branch behind `-Xsource`. It is pretty esoteric so I don't think the demand will be that high for such a backport.

adriaanm commented:
Agreed, 2.12.0-M1 makes more sense.

adriaanm commented:
I'm blown away by the simplification in `isInPackageObject`. Can we get a second opinion on that specific part? /cc @lrytz

retronym commented:
The thinking is: if we are selecting a non-top-level symbol from a package prefix, it could _only_ be from a package object. Before, we used to actually check the members of the corresponding package object to check if the symbol was among them, and only otherwise be convinced we could use the package prefix.

Remember: from the perspective of `Contexts#lookupSymbol`, the members of a package object _are_ members of the package itself, the are in _both_ scopes:

```
scala> ScalaPackageClass.info.decl(TypeName("List"))
res1: $r.intp.global.Symbol = type List

scala> ScalaPackageClass.info.member(nme.PACKAGE).info.decl(TypeName("List"))
res2: $r.intp.global.Symbol = type List
```

`makeAccessible` is responsible for turning the fiction (methods owned by a package) into fact (redirection through the package object).

lrytz commented:
It really looks almost magic, but really great. So LGTM.

We could have a sanity-check: when something is entered into a package scope, its owner should be either the package class, or the package object module class. To be safe, it could be a warning instead of an assertion.

retronym commented:
> its owner should be either the package class, or the package object module class.  

Not quite: it could also be owned by a base class of the package object, which is actually the crux of this PR before it went on a refactoring spree. :smile:

lrytz commented:
true.. :)
